Plans Grid: add monthly pricing option

ciampo opened the pull request:
### Changes proposed in this Pull Request

* Add the `<PlansIntervalToggle>` component to the Plans Grid UI
* Add various UI elements to the Plans Grid (both table and accordion variations) related to monthly plans
* Rewrite the Feature List component, in order to allow any feature to be showing the "requires annually plan" badge
* Hook the UI to the plans data-store, using the billing period selected by the user (monthly or annually) and update the plans accordion / table accordingly
* Updated Step by Step launch flow, so that plan info is displayed correctly in the final checkout step. When the monthly plan is selected, the cost of the premium domain is displayed in the final step too (plus minor style fixes)
* Update the refund window (14 or 7 days) in the plans step in Gutenboarding, based on the value picked by the user when interacting with the plans grid toggle

### Testing instructions

#### Setup

#### Test in Gutenboarding

- Pull this PR's branch on your local machine or test using the live link: https://calypso.live/new?branch=feat/plans-grid/monthly-prices
- If Run `yarn && yarn start` in the root folder of this repo
- Visit `calyspo.localhost:3000/new` and navigate to the Plan step
- Make sure that all A/Cs from #48760 are satisfied (except the follow up/related issues)

#### Test in Focused Launch Flow

- Prepare your sandbox:
  - Make sure you have an active ssh connection to your sandbox
  - Clear your sandbox and pull the latest version from trunk
  - Add an unlaunched site **created via `/start`** to your sandbox
- Pull this PR's branch on your local machine and run the project locally:
  - Run `yarn && yarn start` in the root folder
  - After the command above has finished compiling, in a separate terminal window run `cd apps/editing-toolkit && yarn dev --sync`
- Visit `calyspo.localhost:3000/page/UNLAUNCHED_SITE_CREATED_WITH_START.wordpress.com/home?flags=create/focused-launch-flow`
- Click on the "Launch" button to open Focused Launch flow
- Click on the "View all Plans" button to navigate to the Plan Details step
- Make sure that all A/Cs from #48760 are satisfied (except the follow up/related issues)


#### Test in Step By Step Launch Flow

- Prepare your sandbox:
  - Make sure you have an active ssh connection to your sandbox
  - Clear your sandbox and pull the latest version from trunk
  - Add an unlaunched site **created via `/new`** and **on a free plan** to your sandbox
- Pull this PR's branch on your local machine and run the project locally:
  - Run `yarn && yarn start` in the root folder
  - After the command above has finished compiling, in a separate terminal window run `cd apps/editing-toolkit && yarn dev --sync`
- Visit `calyspo.localhost:3000/page/UNLAUNCHED_SITE_CREATED_WITH_NEW .wordpress.com/home`
- Click on the "Launch" button to open Step By Step Launch flow
- Proceed to the plans step
- Make sure that all A/Cs from #48760 are satisfied (except the follow up/related issues)
- Make sure that checkout step looks correct (all of the listed features are correctly included in the selected plan, including a custom domain in case of annually-billed plans)

Fixes #48760
Fixes #49027

ciampo committed:
Add <PlansIntervalToggle> toggle and billingInterval state to Plans Grid package

ciampo committed:
PlansIntervalToggle props from Type to Interface

ciampo committed:
Rename maxAnnuallySavingPerc to maxMonhtlyDiscountPercentage

ciampo committed:
Add billingInterval and onMaxMonhtlyDiscountPercentageChange props

Added to both variations of the grid (accordion AND table)

ciampo committed:
Feature List: Show nudge next to custom domain feature

ciampo committed:
Update billing details (price note) label

ciampo committed:
Add price discount label below plan price

ciampo commented:
Update on the progress so far:

![image](https://user-images.githubusercontent.com/1083581/104777478-01a3b800-577c-11eb-8496-c98027882b0a.png)
![image](https://user-images.githubusercontent.com/1083581/104777491-09635c80-577c-11eb-9ced-3b21f5de2a9b.png)

**Done**

- Added the Monthly / Annually toggle component to the plans grid
- Mocked some missing data from data-store 
- Started work on implementing UI changes to the accordion version of the plans grid:
  - Changed the billing information below the price to _Per month, billed monthly_ / _Per month, billed as [PRICE] annually_
  - Added the _SAVE [PERC]% BY PAYING ANNUALLY_ text below the plan price (marked with number 1 in the screenshot above)
  - Added the _INCLUDED WITH ANNUAL PLANS_ text above the domain item in the feature list (marked with number 2 and 5 in the screenshot above). I took inspiration from how this was implemented in `/start/plans`, since it was missing from the design specs.

Progress proceeded more slowly than expected, as some parts of the UI had to be refactored to allow for different outcomes depending on the billing period (monthly vs annually)

**What's left**

- [x] Integration with plans data store (once #48790 gets merged). In particular, we need to pull the correct data about prices, discounts (per plan), and maximum discount value.
- [x] Apply same UI changes to the "Table" variant of the plans-grid.
- [x] It also looks like the plans data from the data-store doesn't include, for annually billed plans, the total cost of the plan for one year. Currently in the UI I'm using the monthly cost because I couldn't find the annual cost (see number 4 in the screenshot above).
- [x] By looking at `/start/plans`, it seems that the "Live chat support" feature is also not included with the monthly plans. From a quick look at the plans data-store, it looks like we don't have a way of knowing if a certain feature is not included with the monthly version of a plan. Furthermore, we'd also need to slightly change the way this data is passed down to the UI, since at the moment features are just an array of strings.
- [x] The plan details component, which shows a detailed comparison of all plans feature-by-feature, also needs to be updated reflecting that **Custom domain** (starting from Personal plan) and **Live chat support** (starting from Premium plan) are not included in the monthly version of a paid plan.
- [x] Finally, a note about the feature list layout — by adding the _INCLUDED WITH ANNUAL PLANS_, the spacing between the feature list items becomes irregular, since domain item now takes more vertical space (see number 3 in the screenshot above). One way of tackling this issue would be to set `align-self: end;` on each child element on the CSS grid


cc @razvanpapadopol

ciampo committed:
Table variant: pass billingInterval, onMaxMonhtlyDiscountPercentageChange and mock discount

a8ci18n commented on packages/plans-grid/src/plans-accordion-item/index.tsx:
```diff
@@ -112,17 +118,42 @@ const PlanItem: React.FunctionComponent< Props > = ( {
+								{ ! isFree &&
+									( billingInterval === 'ANNUALLY'
+										? createInterpolateElement(
+												__( 'per month, billed as <price /> annually', __i18n_text_domain__ ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **28** times:
`translate( 'per month, billed annually' )` *ES Score: 8*

a8ci18n commented on packages/plans-grid/src/plans-feature-list/index.tsx:
```diff
@@ -27,60 +28,94 @@ const ChevronDown = (
+		aria-label={
+			billingInterval === 'ANNUALLY'
+				? __( 'Included with annual plans', __i18n_text_domain__ )
+				: __( 'Only included with annual plans', __i18n_text_domain__ )
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Included with annual plans' )` *ES Score: 9*

ciampo commented:
Rebased on top of latest `trunk` and solved conflicts

ciampo committed:
Fix feature list items alignment

ciampo committed:
Refactor plans feature list

- clean the code and use same markup structure for domain button and
  rest of the features
- align to latest designs
- all features can display the "available only if annual plan" badge
- mock requiresAnnuallyBilledPlan on features array

ciampo committed:
Orange text and orange cross icon for "annually-only" disabled features

ciampo committed:
Darker green for SAVE X% text

ciampo committed:
Use sprintf instead of createInterpolateElement

ciampo committed:
Temporarily disable ESLint for ts-ignore exceptions

ciampo committed:
Tweak feature list design

- Darker green
- Different margin-top depending on the design variant

ciampo committed:
Bing savings discount percentage label to "plans-table" variation

a8ci18n commented on packages/plans-grid/src/plans-table/plan-item.tsx:
```diff
@@ -134,9 +139,34 @@ const PlanItem: React.FunctionComponent< Props > = ( {
+							{ ! isFree &&
+								( billingInterval === 'ANNUALLY'
+									? sprintf(
+											__( 'per month, billed as %s annually', __i18n_text_domain__ ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **3** times:
`translate( 'per month{{br/}}billed yearly' )` *ES Score: 7*

a8ci18n commented on packages/plans-grid/src/plans-table/plan-item.tsx:
```diff
@@ -134,9 +139,34 @@ const PlanItem: React.FunctionComponent< Props > = ( {
+						>
+							{ sprintf(
+								// Translators: "%s" is a number, and "%%" is the percent sign. Please keep the "%s%%" string unchanged when translating.
+								__( 'Save %s%% by paying annually', __i18n_text_domain__ ),
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( 'Save %(discountRate)s%% by paying annually' )` *ES Score: 8*

ciampo committed:
Update plan item comment

ciampo committed:
Mark features that require the annual version of a plan

- added hardcoded list of feature ids to plans data-store
- added requiresAnnuallyBilledPlan flag to "features" in plans data-store
- compute value for each feature plan's feature directly in the UI
  (plans-feature-list component)

ciampo committed:
Plan comparison table: Support for features that require yearly plan

ciampo committed:
Match darker green color for feature list ticks

ciampo committed:
Stronger CSS override for segmented control

ciampo commented:
Hey @ollierozdarz , we’ve been working on the UI changes to enable monthly plans in Gutenboarding (and Focused Launch and Step by Step Launch).

So far we followed the “Onboarding” and “Launch flow” design specs on Figma, but in a few cases we had to go with our own judgment — mainly because of some features that are supported in the reskinned signup & launch flows.

One thing to keep in mind is that the “Plans Grid” package is re-used across different flows (gutenboarding, step by step launch, and focused launch), and so _the changes that we make are going to be applied across all these flows_.

**We would like you to visit the [staging link](https://hash-6bbdaeb99d4abfb55f207347f167b495f5f5d873.calypso.live/new)  and give some early feedback on this version of the plans grid in Gutenboarding.**

In particular, these are some items that we’d like your feedback on:
- We’ve added the “INCLUDED WITH ANNUAL PLANS” text above all features that require the annual version of a plan. We followed the implementation from the reskinned signup/launch flows, but we changed a few details in order to match more closely with the style of this new UI (e.g. we’re using orange instead of red for the text, we’re striking through the original grey text, and we’re showing an orange cross icon when the feature is not included)
- We added the same “INCLUDED WITH ANNUAL PLANS” text also in the comparison table at the bottom of the page. When the monthly version of a plan is selected, the ticks are turned into crosses
- We changed the “SAVE X%” text to  “SAVE X% BY PAYING ANNUALLY” to achieve feature parity with the reskinned signup/launch flows
- The feature items in each plan’s card have green ticks, although we changed the green to be the same as the “SAVE X% BY PAYING ANNUALLY” text
- Following the latest Figma specs, the interactive “custom domain” item (the first in the list of features) has dark grey text. This is different from what’s currently in production (that text is blue instead of gray)

As a caveat, please keep in mind that at the moment the data for monthly plans has not been integrated yet, so please ignore the figures (prices and discounts) about monthly plans during your review (you can actually see the complete list of TODO items at the [top of this page](https://github.com/Automattic/wp-calypso/pull/48963#issue-555788450)).

Here's also a screen capture, in case the staging link doesn't work for you:


https://user-images.githubusercontent.com/1083581/105221846-f674d180-5b59-11eb-9b9e-739dd3674d05.mp4



cc @razvanpapadopol

ollierozdarz commented:
Hi @ciampo, thanks for the ping. First of all, great job on this! 👏

Rick has asked me to take a look at giving this specific page some more love in the near future (specifically the plans accordion), so I'm not keen to change much here now, and I think this is nearly at a state where it's at parity to ship.

I did notice a few small things that could be improved and listed them below: 

<img width="348" alt="Screen Shot 2021-01-21 at 12 03 12 pm" src="https://user-images.githubusercontent.com/67085494/105270106-ed7b1300-5be0-11eb-9996-82cc1e34f3a4.png">

<img width="486" alt="Screen Shot 2021-01-21 at 12 03 16 pm" src="https://user-images.githubusercontent.com/67085494/105270125-f966d500-5be0-11eb-8516-dd2fc028f16d.png">

<img width="460" alt="Screen Shot 2021-01-21 at 12 03 23 pm" src="https://user-images.githubusercontent.com/67085494/105270136-fe2b8900-5be0-11eb-8821-9bb4be32738a.png">

For the note above, I never specified this in my refined reskin onboarding Figma files (my bad! 🙂), but I think when this link is active we should underline it to make it stand out more (i.e. the user is able to select a URL since they have Annual toggled and the correct plan).

<img width="1505" alt="Screen Shot 2021-01-21 at 12 07 48 pm" src="https://user-images.githubusercontent.com/67085494/105270324-4fd41380-5be1-11eb-9dfa-1bb805bfa9ef.png">

For the **After** screenshot, I've just aligned the checks and crosses with the feature label (for consistency) and added more bottom margin to the 'Included with annual plans', since they were sitting very close to the text below. While we're here, we may as well fix the border-radius for the primary container (it should be using the same [looks like 5px] on all four corners).

paulbonahora commented:
Great work on this! I'd like to add some further feedback:

- Domain selection step: replace "included in plans" for "included on annual plans"
- Refund policy at the top of plans selection page: annual plans support 30 days ([soon to be 14 days](https://growthchannels.wordpress.com/2020/12/23/project-thread-14-days-refund-marketing-pt/#comment-2218), we'll have to update this on Friday so it sounds like it's good timing to refine this - didn't have this on my radar as part of the 14 days project). Refund policy for monthly plans is 7 days. The message could change depending on whether the user selects monthly or yearly, just to be clearer to the user: "There is no risk. You can cancel your annual plan for a full refund within 14 days" or "There is no risk. You can cancel your monthly plan for a full refund within 7 days".
- Discounts shown for annual plans (the 43%) will vary slightly per currency, just as an FYI.

ciampo commented:
Thank you @ollierozdarz and @paulbonahora for your feedback!

Here's some quick notes in response to @ollierozdarz 's feedback:
- The border radius and the small dark outline in the "Annually"/"Monthly" toggle are some weird glitches that are not supposed to be there. I'll have a look and fix those
- Noted re. underlining the "domain" link when it's active, and using a darker shade of grey
- Noted re. aligning the check/cross with the feature label
- Noted re. adding more bottom margin to the "'Included with annual plans'" text
- Noted re. increasing the border-radius of the plan containers to 5px

And here's some quick notes in response to @paulbonahora 's feedback:
- Noted re. replacing "included in plans" for "included on annual plans"
- Noted re. rewording the refund policy at the top of the plans page
- Regarding the discount figures (and other data about the monthly plans), please note that this PR is still WIP and that those figures at the moment are just placeholders. They will likely be updated today and use real-world, dynamic data form the APIs

ciampo committed:
Fix more segmented control style leaks with stringer CSS selectors

ciampo committed:
Design feedback around "custom-domain" feature button:

- added back is-cta class
- when is-cta, use the darkest grey and underline the text
- when it's disabled (for whatever reason), strikethrough
- added cursor pointer

ciampo committed:
Design feedback: Align cross/tick icon at bottom of each feature item

ciampo committed:
Design feedback: add more margin bottom under "INCLUDED WITH ANNUAL PLANS" text

ciampo committed:
Fix text not breaking the line correctly in the feature list

ciampo committed:
Design feedback: add more margin bottom under "INCLUDED WITH ANNUAL PLANS" text in comparison table

ciampo committed:
Design feedback: fix border radius of first accordion item

ciampo committed:
Fix getting plan's price after rebase

ciampo committed:
Add PlanBillingPeriod type to data-store, and re-use it across the plans-grid package

- delete useBillingPeriod hook
- delete BillingIntervalType
- rename all `billingInterval` vars to `billingPeriod`

ciampo committed:
Use annual price for plans annual billing cost

ciampo committed:
Get real annual discount percentage from data-store

ciampo committed:
Compute max annual discount dynamically

ciampo committed:
Fix table-variant bugs:

- fallback to 0 when discount value is undefined (for free plan)
- add visibility:hidden for when that text is marked as hidden

ciampo commented:
Update:

- all design feedback has been addressed
- rebased on top of trunk after #48790 got merged
- started integration work with data-store

ciampo committed:
Fix unit tests

ciampo committed:
Remove @ts-ignore, not necessary after last dependencies update

ciampo committed:
Aggregate all plans store selectors into one useSelect function call

ciampo committed:
Move features computation to data-store, add sorting

Features that required the annually billed version of a plan are first

ciampo commented:
Update:
- rebased on top of latest trunk
- refactored some data-store related code:
  - aggregated selectors into same `useSelect` hook call
  - moved features data processing to data store
  - added features sorting, so that features that require the annually billed version of a plan are first

ciampo committed:
Fix unit tests

ciampo committed:
Feature list: add more margin top for features that require annual plans

ciampo committed:
Gutenboarding: change refund window days in plans step (7 vs 14)

ciampo committed:
Gutenboarding: refactor plans step useSelect calls

ciampo commented:
Update:
- Rebased in order to include changes from  #48872 and #49168
- Worked on the plans step in Gutenboarding — the days of the refund window now update based on the monthly vs annually toggle
- Started creating follow-up issue based on previous feedback and TODO items

ciampo committed:
Taller margin before actions button in accordion cars

ciampo committed:
Step by Step flow: display plan product info after change to monthly/annually

- refactor all calls to LaunchStore into one useSelect
- get the plan product from the PlanStore, useful to get the slug and billingPeriod
- show only the effective features that come with the selected plan
- fix tick icon sizing (it would shrink on small viewports)

ciampo committed:
Set a max-width for the "SAVE XX%..." text

This prevents that text from overlapping with action buttons in step by step flow

ciampo commented:
Update:

- Fixed code to retrieve the selected plan product in the Step by Step flow (checkout step)
  - also updated how features are displayed
  - also fixed some minor CSS glitches
- Reported [this issue](https://github.com/Automattic/wp-calypso/issues/48760#issuecomment-766952241) in the Step by Step final (checkout) step — we can start reviewing this PR while we agree on the best way to get around it

ciampo committed:
Add TODO item in Step by Step launch (checkout step)

ciampo commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx:
```diff
@@ -59,6 +72,7 @@ const FinalStep: React.FunctionComponent< LaunchStepProps > = ( { onNextStep, on
 
 	const domainSummary = (
 		<div className="nux-launch__summary-item">
+			{ /* TODO: what if the user picks a monthly-billed plan?  */ }
```
See https://github.com/Automattic/wp-calypso/issues/48760#issuecomment-766952241

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/a4e89e697398c082479c44f948d5604d081cc501

alshakero commented on packages/data-stores/src/plans/resolvers.ts:
```diff
@@ -95,11 +98,59 @@ function processFeatures( features: Feature[] ) {
+	);
+
+	// Features requiring an annually billed plan should be first in the array.
+	features.sort( ( a, b ) => {
+		if ( a.requiresAnnuallyBilledPlan === b.requiresAnnuallyBilledPlan ) {
+			// Either they both require an annually billed plan, or they both don't
+			return 0;
+		} else if ( a.requiresAnnuallyBilledPlan ) {
+			// "a" requires an annually billed plan, "b" does not
+			return -1;
+		}
+
+		// By exclusion, "b" requires an annually billed plan, "a" does not
+		return 1;
+	} );
+
+	return features;
+}
```
It might be less readable, but this is the same as: 

```suggestion
    // Features requiring an annually billed plan should be first in the array.
	features.sort(
		( a, b ) => Number( a.requiresAnnuallyBilledPlan ) - Number( b.requiresAnnuallyBilledPlan )
	);
```

ciampo replied:
I personally prefer code that is easy to understand, but I quite like how neat this looks!

ciampo replied:
Turns out it had to be the other way around (`b - a`)

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/db937b810c139622c334991288de4967ccbd6073

alshakero commented on packages/data-stores/src/plans/resolvers.ts:
```diff
@@ -95,11 +98,59 @@ function processFeatures( features: Feature[] ) {
 	}, {} as Record< string, PlanFeature > );
 }
 
+function doesFeatureRequireAnnuallyBilledPlan(
```
I love verbosity, but this is too verbose IMO, `requiresAnnual` has all the information you need when combined with the parameters names (which the user knows thanks to TypeScript).

ciampo replied:
I'm going to rename it "featureRequiresAnnual" to still give the notion that it's about features

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/671cb739aa44c4f9a3230acad3fb33b0c5d6c714

alshakero commented on packages/plans-grid/src/plans-grid/index.tsx:
```diff
@@ -63,15 +65,37 @@ const PlansGrid: React.FunctionComponent< Props > = ( {
+		}
+	}, [ billingPeriod, onBillingPeriodChange ] );
+
+	const [ maxMonhtlyDiscountPercentage, setMaxMonhtlyDiscountPercentage ] = React.useState<
+		number | undefined
+	>( undefined );
```
Can't we use [the hook](useSupportedPlans) instead of this?

ciampo replied:
The link to the hook is broken for me, can you double check?

alshakero replied:
Sure! I mean this [hook ](https://github.com/Automattic/wp-calypso/pull/48963/files#diff-3242762916fde7013a4cb5ab3f61bfa25b101956ed6b517e9a2ee92b8b009e52R12).

ciampo replied:
Thank you! Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/e2bff7c156d04a06b3983da50e2e638a11fb7e79

alshakero commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -86,15 +83,15 @@ const PlansIntervalToggle: React.FunctionComponent< ToggleHostProps > = ( {
 						{ __( 'Annually', __i18n_text_domain__ ) }
 					</span>
-					{ intervalType === 'MONTHLY' && (
+					{ intervalType === 'MONTHLY' && typeof maxMonhtlyDiscountPercentage !== undefined && (
```
```suggestion
					{ intervalType === 'MONTHLY' && maxMonhtlyDiscountPercentage && (
```

ciampo replied:
`maxMonhtlyDiscountPercentage` could technically be `0` and fail this check  — I think that checking explicitly for it to be undefined is better in this instance

alshakero replied:
Good point but would we want to show "save up to 0%"?

ciampo replied:
That's also a good point. I'll leave it up to you — I'm good with both options. I'll just add a comment in any case

alshakero replied:
It's a nit anyways. I'd go for `&& maxMonhtlyDiscountPercentage` because it covers two cases (0 and undefined).

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/c71139b0fc0d0f20421f1dbca88a1a327acaa187

alshakero commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -46,17 +43,17 @@ export const PopupMessages: React.FunctionComponent = ( { children } ) => {
 	onChange,
 	intervalType,
-	maxSavingsPerc,
+	maxMonhtlyDiscountPercentage,
```
```suggestion
	maxMonthlyDiscountPercentage,
```

ciampo replied:
It took be 2 minutes of staring at this suggestion to understand the typo 🤦  I'll have to refactor the code in a few places — nice catch!

alshakero replied:
I highly recommend [this](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)! Can save you from bugs too (in JS). 

![image](https://user-images.githubusercontent.com/17054134/105857515-a9c24800-5fea-11eb-94e9-e672ea0c9938.png)

ciampo replied:
🙏

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/146a5b7927b5093cb2803108bba8dee4bf853365

alshakero commented on packages/plans-grid/src/plans-table/index.tsx:
```diff
@@ -43,15 +44,19 @@ const PlansTable: React.FunctionComponent< Props > = ( {
-	);
+	const { supportedPlans, maxAnnualDiscount } = useSupportedPlans( locale, billingPeriod );
+
+	React.useEffect( () => {
+		onMaxMonhtlyDiscountPercentageChange( maxAnnualDiscount );
+	}, [ onMaxMonhtlyDiscountPercentageChange, maxAnnualDiscount ] );
```
If I understand correctly, if you use the hook as I commented above, we won't need this.

alshakero reviewed (changes requested):
Great work on this! This is almost ready, I'm just waiting for confirmation on `setMaxMonhtlyDiscountPercentage`.

Also, not sure if this is in scope, but when you re-open the plans grid, the period selection is reset to annual, it would be nice if we persisted this.

ciampo commented:
> but when you re-open the plans grid, the period selection is reset to annual, it would be nice if we persisted this.

I agree, but I didn't implement it because it was not in scope

@razvanpapadopol , what do you you think? Should we open a follow-up issue for persisting the toggle selection in the launch store?

ciampo commented on packages/data-stores/src/plans/constants.ts:
```diff
@@ -184,3 +184,9 @@ export const currenciesFormats: Record< string, Currency > = {
+export const FEATURE_IDS_THAT_REQUIRE_ANNUALLY_BILLED_PLAN = [
+	'custom-domain',
+	'support-live',
+	// 'priority-support',
```
TODO: potentially add new slugs after #49142 is closed

ciampo committed:
Add FEATURES_IDS_THAT_REQUIRE_ANNUALLY_BILLED_PLAN variable back

It was deleted by mistake during a rebase

ciampo committed:
Refactor features sorting in a more concise way

ciampo committed:
Refactor maxMonthlyDiscountPercentage + typos and function renaming

Rename "doesFeatureRequireAnnuallyBilledPlan" to "featureRequiresAnnual"

Fix typo: Monhtly => Monthly

Refactor maxMonthlyDiscountPercentage truthy check (+ typo)

Refactor maxMonthlyDiscountPercentage truthy check

Typo

alshakero reviewed (commented):
In case the hook comment isn't clear, here is a patch of what I mean:

```diff
diff --git a/packages/plans-grid/src/plans-accordion/index.tsx b/packages/plans-grid/src/plans-accordion/index.tsx
index 20b09dfd11..9a75e21bf6 100644
--- a/packages/plans-grid/src/plans-accordion/index.tsx
+++ b/packages/plans-grid/src/plans-accordion/index.tsx
@@ -39,7 +39,6 @@ export interface Props {
 	disabledPlans?: DisabledPlansMap;
 	locale: string;
 	billingPeriod: Plans.PlanBillingPeriod;
-	onMaxMonthlyDiscountPercentageChange: ( perc: number | undefined ) => void;
 }
 
 const PlansAccordion: React.FunctionComponent< Props > = ( {
@@ -51,15 +50,10 @@ const PlansAccordion: React.FunctionComponent< Props > = ( {
 	disabledPlans,
 	locale,
 	billingPeriod,
-	onMaxMonthlyDiscountPercentageChange,
 } ) => {
 	const { __ } = useI18n();
 
-	const { supportedPlans, maxAnnualDiscount } = useSupportedPlans( locale, billingPeriod );
-
-	React.useEffect( () => {
-		onMaxMonthlyDiscountPercentageChange( maxAnnualDiscount );
-	}, [ onMaxMonthlyDiscountPercentageChange, maxAnnualDiscount ] );
+	const { supportedPlans } = useSupportedPlans( locale, billingPeriod );
 
 	const isLoading = ! supportedPlans?.length;
 	const placeholderPlans = [ 1, 2, 3, 4 ];
diff --git a/packages/plans-grid/src/plans-grid/index.tsx b/packages/plans-grid/src/plans-grid/index.tsx
index 0551e2a116..b6fb090222 100644
--- a/packages/plans-grid/src/plans-grid/index.tsx
+++ b/packages/plans-grid/src/plans-grid/index.tsx
@@ -20,6 +20,7 @@ import type {
 	PopularBadgeVariation,
 } from '../plans-table/types';
 import PlansIntervalToggle from '../plans-interval-toggle';
+import { useSupportedPlans } from '../hooks';
 
 /**
  * Style dependencies
@@ -79,9 +80,7 @@ const PlansGrid: React.FunctionComponent< Props > = ( {
 		}
 	}, [ billingPeriod, onBillingPeriodChange ] );
 
-	const [ maxMonthlyDiscountPercentage, setMaxMonthlyDiscountPercentage ] = React.useState<
-		number | undefined
-	>( undefined );
+	const { maxAnnualDiscount } = useSupportedPlans( locale, billingPeriod );
 
 	isAccordion && debug( 'PlansGrid accordion version is active' );
 
@@ -92,7 +91,7 @@ const PlansGrid: React.FunctionComponent< Props > = ( {
 			<PlansIntervalToggle
 				intervalType={ billingPeriod }
 				onChange={ setBillingPeriod }
-				maxMonthlyDiscountPercentage={ maxMonthlyDiscountPercentage }
+				maxMonthlyDiscountPercentage={ maxAnnualDiscount }
 				className="plans-grid__toggle"
 			/>
 
@@ -108,7 +107,6 @@ const PlansGrid: React.FunctionComponent< Props > = ( {
 							disabledPlans={ disabledPlans }
 							locale={ locale }
 							billingPeriod={ billingPeriod }
-							onMaxMonthlyDiscountPercentageChange={ setMaxMonthlyDiscountPercentage }
 						></PlansAccordion>
 					) : (
 						<PlansTable
@@ -124,7 +122,6 @@ const PlansGrid: React.FunctionComponent< Props > = ( {
 							showTaglines={ showPlanTaglines }
 							defaultAllPlansExpanded={ defaultAllPlansExpanded }
 							billingPeriod={ billingPeriod }
-							onMaxMonthlyDiscountPercentageChange={ setMaxMonthlyDiscountPercentage }
 						></PlansTable>
 					) }
 				</div>
diff --git a/packages/plans-grid/src/plans-table/index.tsx b/packages/plans-grid/src/plans-table/index.tsx
index e214433ff7..bab58eb73f 100644
--- a/packages/plans-grid/src/plans-table/index.tsx
+++ b/packages/plans-grid/src/plans-table/index.tsx
@@ -34,7 +34,6 @@ export interface Props {
 	customTagLines?: CustomTagLinesMap;
 	defaultAllPlansExpanded?: boolean;
 	billingPeriod: Plans.PlanBillingPeriod;
-	onMaxMonthlyDiscountPercentageChange: ( perc: number | undefined ) => void;
 }
 
 const PlansTable: React.FunctionComponent< Props > = ( {
@@ -45,18 +44,13 @@ const PlansTable: React.FunctionComponent< Props > = ( {
 	disabledPlans,
 	locale,
 	billingPeriod,
-	onMaxMonthlyDiscountPercentageChange,
 	showTaglines = false,
 	CTAVariation = 'NORMAL',
 	popularBadgeVariation = 'ON_TOP',
 	customTagLines,
 	defaultAllPlansExpanded = false,
 } ) => {
-	const { supportedPlans, maxAnnualDiscount } = useSupportedPlans( locale, billingPeriod );
-
-	React.useEffect( () => {
-		onMaxMonthlyDiscountPercentageChange( maxAnnualDiscount );
-	}, [ onMaxMonthlyDiscountPercentageChange, maxAnnualDiscount ] );
+	const { supportedPlans } = useSupportedPlans( locale, billingPeriod );
 
 	const [ allPlansExpanded, setAllPlansExpanded ] = useState( defaultAllPlansExpanded );
 
```

ciampo committed:
Use the useSupportedPlans instead of callback props hell to get maxAnnualDiscount

ciampo committed:
Step by Step: show domain price when monthly-billed plan is selected

ciampo commented:
Rebased on top of latest trunk.

Hey @alshakero , all feedback has been addressed apart from [this comment](https://github.com/Automattic/wp-calypso/pull/48963#discussion_r564681643) (which is waiting for D55944-code to be approved/deployed).

Can you give it another round of review?

razvanpapadopol commented on packages/plans-grid/src/plans-feature-list/style.scss:
```diff
@@ -22,123 +30,159 @@ ul.plans-feature-list__item-group {
+
 .plans-feature-list__placeholder {
 	@include onboarding-placeholder();
 	display: inline-block;
```
We need more spacing between column on feature list. The overlapping is visible on languages with longer texts or on smaller screen sizes.

Probably related to the changes in this file.

<img width="802" alt="Screenshot 2021-01-27 at 12 40 56" src="https://user-images.githubusercontent.com/14192054/105980785-f61a9000-609d-11eb-8122-4575ed3b8560.png">

ciampo replied:
Good catch! Should be fixed in https://github.com/Automattic/wp-calypso/pull/48963/commits/ba30e28e653c7227085cdf0d32ae4b9b6b441a72 

With verbose languages, the feature layout doesn't always look pretty, but I don't think it should be a blocker for this PR
<img width="967" alt="Screenshot 2021-01-27 at 13 14 44" src="https://user-images.githubusercontent.com/1083581/105989584-a0e07d80-60a1-11eb-9066-443290e7c347.png">

ciampo committed:
Make sure that feature text doesn't overflow the parent's width

alshakero reviewed (changes requested):
It seems this is breaking focused launch. I'm getting a fatal exception (see [here](https://reactjs.org/docs/error-decoder.html/?invariant=31&args[]=object%20with%20keys%20%7BdisabledLabel%7D&args[]=)).  

I think it only happens if you have a custom domain.

The fix seems to be this:
```diff
diff --git a/packages/plans-grid/src/plans-feature-list/index.tsx b/packages/plans-grid/src/plans-feature-list/index.tsx
index 48983e9cfa..13f2e322d8 100644
--- a/packages/plans-grid/src/plans-feature-list/index.tsx
+++ b/packages/plans-grid/src/plans-feature-list/index.tsx
@@ -205,7 +205,7 @@ const PlansFeatureList: React.FunctionComponent< PlansFeatureListProps > = ( {
 	if ( disabledLabel ) {
 		featureItems.push( {
 			bulletIcon: CrossIcon,
-			textNode: { disabledLabel },
+			textNode: disabledLabel,
 			requiresAnnuallyBilledPlan: false,
 			listItemExtraClassName: 'plans-feature-list__item--disabled-message',
 		} );
```

ciampo committed:
Fix: pass string instead of { string } as React.Node

Fix: wrap { disabledLabel > inside a React fragment to avoid render errors

Use strings as text nodes instead of react fragments

ciampo commented:
> It seems this is breaking focused launch. I'm getting a fatal exception (see [here](https://reactjs.org/docs/error-decoder.html/?invariant=31&args%5B%5D=object%20with%20keys%20%7BdisabledLabel%7D&args%5B%5D=)).

Good catch — I fixed it in https://github.com/Automattic/wp-calypso/pull/48963/commits/2204df97714632802a3611c6aa4b610688040c75 and https://github.com/Automattic/wp-calypso/pull/48963/commits/ad4c8f1f38a278dbe46845254ef5f2ac7f8e2d7c

alshakero reviewed (approved):
1. Tested every item in the A/C in Gutenboarding. LGTM!
2. Smoke tested focused launch. The summary still shows annual prices when picking a monthly plan. Maybe we need to add [this](https://github.com/Automattic/wp-calypso/issues/49257) to the list of [standing issues](https://github.com/Automattic/wp-calypso/issues/48760#:~:text=Follow%2Dup%2Frelated%20issues%3A).
3. Smoke, but extensively tested step by step launch. Works like a charm.
4. Reviewed the code a couple times. LGTM!

ciampo commented:
Oh hold until #49142 is closed and the feature slug changes are applied to this PR

ciampo committed:
Update IDs of features requiring annually billed plans

razvanpapadopol commented:
Found some i18n issues. Here is a comparison with reskinned:

### Monthly plans:
* "monthly" translation is not the same
* "annually" is not translated
* "Save xx% by paying anually" is not translated

#### `/new` on live:
<img width="1235" alt="Screenshot 2021-01-27 at 16 40 14" src="https://user-images.githubusercontent.com/14192054/106007625-2cb4d280-60bf-11eb-991c-7d3beec49272.png">

#### `/start` in production:
<img width="1168" alt="Screenshot 2021-01-27 at 16 44 32" src="https://user-images.githubusercontent.com/14192054/106007652-32121d00-60bf-11eb-8e5a-25c30780103d.png">

### Yearly plans:
* _/month_ and _per month_ seems like an unnecessary repetition
* _per month, billed as..._ is not translated

#### `/new` on live :
<img width="738" alt="Screenshot 2021-01-27 at 16 43 44" src="https://user-images.githubusercontent.com/14192054/106007681-39d1c180-60bf-11eb-8e1c-6a98b3146e18.png">

#### `/new` in production:
<img width="1303" alt="Screenshot 2021-01-27 at 16 50 34" src="https://user-images.githubusercontent.com/14192054/106008339-dbf1a980-60bf-11eb-9ec0-db9f6f04d398.png">

#### `/start` in production :
<img width="520" alt="Screenshot 2021-01-27 at 16 48 23" src="https://user-images.githubusercontent.com/14192054/106007966-87e6c500-60bf-11eb-802f-8eb595af5f17.png">

ciampo commented:
Now that #49142 has been solved, I updated the feature IDs in https://github.com/Automattic/wp-calypso/pull/48963/commits/d7dbdea04e8a50b6536ddf6fcaeba5d23f887e2f

This is how things look at the moment:

![image](https://user-images.githubusercontent.com/1083581/106009387-8ebe0980-60b8-11eb-954c-1aa125ea1d09.png)

![image](https://user-images.githubusercontent.com/1083581/106009395-92519080-60b8-11eb-892c-5c32aa5e7e31.png)

![image](https://user-images.githubusercontent.com/1083581/106009403-97164480-60b8-11eb-99f8-08646e466a60.png)

@razvanpapadopol does this look correct to you, or are there any other feature slugs that need to be added/removed to the list?

ciampo committed:
Fix: use strings with matching translations  (see details in commit log)

Fix translations: use same strings as /start for plans toggle

Fix translations: use same strings as /start form "SAVE X% BY PAYING ANUALLY"

Add "Domain Registration" string in step by step launch final step

Fix translations: "per month, billed as %s annually" => "billed annually"

This is because there are already translations matching what's in /start

Fix translations: use double percent sign to avoid sprintf errors

razvanpapadopol commented on apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/src/launch-steps/final-step/index.tsx:
```diff
@@ -71,6 +71,17 @@ const FinalStep: React.FunctionComponent< LaunchStepProps > = ( { onNextStep, on
+						<>
+							<br />
+							<span className="nux-launch__summary-item__domain-price">
+								{
```
Tiny suggestion:

```suggestion
								{ __( 'Domain Registration', 'full-site-editing' ) }:{ ' ' }
								{
```

**Demo:**
<img width="1112" alt="Screenshot 2021-01-27 at 17 39 16" src="https://user-images.githubusercontent.com/14192054/106014945-a8fee400-60c6-11eb-93e9-a6dff060014b.png">

Ignoring the existing broken i18n, the string should be already translated if we look at:
https://translate.wordpress.com/projects/wpcom/-all-translations/42132/

ciampo replied:
Done in https://github.com/Automattic/wp-calypso/pull/48963/commits/87ce4f05fcb138ef308f915abb3818c6ad28341c

ciampo commented:
Hey @razvanpapadopol , here's my feedback :

---

>"monthly" translation is not the same
>"annually" is not translated

I updated the text in the toggle to "Pay monthly" / "Pay annually", as it is in `/start` — it should appear as correctly translated

---

The text in the tooltip (_Save up to X% by paying annually and get a free domain for one year_) won't be translated for this initial release — the issue has been discussed in detail on p1611765545146500/1611740503.105000-slack-C0KDTA48Y , and it's been decided that it's not a blocker for initial release.

---

> "Save xx% by paying annually" is not translated

I updated that text to be the same as in `/start` — it should appear as correctly translated


---

> per month, billed as... is not translated

I changed that string to "billed annually", as it is in `/start` — it should appear as correctly translated

--- 

> /month and per month seems like an unnecessary repetition

I kept the "per month, billed monthly" string (visible when picking the monthly billed version of a plan) because "billed monthly" doesn't seem to be translatable


--- 

Translations in Step by Step and Focused Launch flow (Editing Toolkit) are not coming through correctly, but I don't think that's an issue related to this PR

alshakero reviewed (approved):
Everything works (is translated) as described.

paulbonahora commented:
Reporting two issues I've seen while testing:

1. https://d.pr/i/oWP8kX
2. https://d.pr/i/KSJy5P

Number 1: we don't offer live chat support on monthly plans. I'd suggest we cross out that item as in the free domain and live chat support whenever monthly billing is selected. Open to suggestions on how to improve the UI for this though!

Number 2: I'd remove the "(billed yearly)". This should happen only when the monthly option is selected.

ciampo committed:
Add priority support to the list of features that required the annual plan

ciampo committed:
Translations: remove (billed yearly) from comparison table if monthly plans are selected

Translations fix: remove "billed yearly" from bottom of details plan

Translations fix: remove (billed yearly) from comparison table only for monthly plans

ciampo commented:
Hey @paulbonahora , I've made a couple of changes based on your feedback:

> Number 1: we don't offer live chat support on monthly plans. 

I went ahead and marked "24/7 Priority live chat support" as a feature that requires the annual version of a plan. Here's who it looks:

| Annually | Monthly |
|---|---|
| <img width="1151" alt="Screenshot 2021-01-27 at 19 13 17" src="https://user-images.githubusercontent.com/1083581/106034855-b79fc800-60d3-11eb-9ef3-deab291a0391.png"> | <img width="1093" alt="Screenshot 2021-01-27 at 19 17 16" src="https://user-images.githubusercontent.com/1083581/106035322-47457680-60d4-11eb-98b6-44e52e7c1732.png"> |

> Number 2: I'd remove the "(billed yearly)"

I removed "(billed yearly)" when the user picks "monthly" from the toggle. Both versions of this text should have translations already available

| Annually | Monthly |
|---|---|
| <img width="1053" alt="Screenshot 2021-01-27 at 19 18 55" src="https://user-images.githubusercontent.com/1083581/106035516-81167d00-60d4-11eb-9c17-44e68cc0e670.png">  |  <img width="1085" alt="Screenshot 2021-01-27 at 19 18 42" src="https://user-images.githubusercontent.com/1083581/106035487-79ef6f00-60d4-11eb-9ec8-f9846b5c33d1.png"> |

razvanpapadopol reviewed (approved):
Tested `/new` in EN and FR and it looks good. I guess we'll re-test the ET before the next version release.
🚢

paulbonahora commented:
It looks like there's confusion with the wording on /new/plans page at the very bottom of the page when the yearly option is selected. See [screenshot here](https://d.pr/i/VDZAN4).

I'd suggest replacing the wording "Monthly Subscription (billed yearly)" to "Monthly Price (billed yearly)" for clarity. There's no change needed when the toggle is set to monthly.

cc: @simison

tyxla commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -84,17 +81,22 @@ const PlansIntervalToggle: React.FunctionComponent< ToggleHostProps > = ( {
 									__i18n_text_domain__
 								),
-								maxSavingsPerc
+								{ maxDiscount: maxMonthlyDiscountPercentage ?? 0 }
```
Seems like we're checking for `maxMonthlyDiscountPercentage` above, so it will never be `null` / `undefined`, therefore there should be no need for the nullish coalescing I think?
